File: Android Car Dealer Tracker/app/src/main/java/com/ezcar24/business/util/calculator/InventoryMetricsCalculator.kt
package com.ezcar24.business.util.calculator

import com.ezcar24.business.data.local.Expense
import com.ezcar24.business.data.local.HoldingCostSettings
import com.ezcar24.business.data.local.InventoryAlert
import com.ezcar24.business.data.local.InventoryAlertType
import com.ezcar24.business.data.local.Vehicle
import com.ezcar24.business.data.local.VehicleInventoryStats
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.Date
import java.util.UUID

object InventoryMetricsCalculator {

    private const val DISPLAY_SCALE = 2

    private const val AGING_BUCKET_0_30 = "0-30"
    private const val AGING_BUCKET_31_60 = "31-60"
    private const val AGING_BUCKET_61_90 = "61-90"
    private const val AGING_BUCKET_90_PLUS = "90+"

    private const val THRESHOLD_AGING_60 = 60
    private const val THRESHOLD_AGING_90 = 90
    private const val THRESHOLD_LOW_ROI = 10
    private const val THRESHOLD_HIGH_HOLDING_COST_PERCENT = 15

    fun calculateAgingBucket(daysInInventory: Int): String {
        return when {
            daysInInventory <= 30 -> AGING_BUCKET_0_30
            daysInInventory <= 60 -> AGING_BUCKET_31_60
            daysInInventory <= 90 -> AGING_BUCKET_61_90
            else -> AGING_BUCKET_90_PLUS
        }
    }

    fun calculateInventoryStats(
        vehicle: Vehicle,
        expenses: List<Expense>,
        settings: HoldingCostSettings
    ): VehicleInventoryStats {
        val daysInInventory = HoldingCostCalculator.calculateDaysInInventory(vehicle)
        val agingBucket = calculateAgingBucket(daysInInventory)

        val improvementExpenses = HoldingCostCalculator.getImprovementExpenses(expenses)
        val holdingCostAccumulated = HoldingCostCalculator.calculateAccumulatedHoldingCost(
            vehicle,
            settings,
            expenses
        )

        val totalCost = VehicleFinancialsCalculator.calculateTotalCost(
            vehicle,
            expenses,
            holdingCostAccumulated
        )

        val roiPercent = vehicle.salePrice?.let { salePrice ->
            VehicleFinancialsCalculator.calculateROI(salePrice, totalCost)
        }

        val profitEstimate = vehicle.askingPrice?.let { askingPrice ->
            VehicleFinancialsCalculator.calculateProfitEstimate(askingPrice, totalCost)
        }

        val now = Date()

        return VehicleInventoryStats(
            id = UUID.randomUUID(),
            vehicleId = vehicle.id,
            daysInInventory = daysInInventory,
            agingBucket = agingBucket,
            totalCost = totalCost,
            holdingCostAccumulated = holdingCostAccumulated,
            roiPercent = roiPercent,
            profitEstimate = profitEstimate,
            lastCalculatedAt = now,
            createdAt = now,
            updatedAt = now
        )
    }

    fun generateInventoryAlerts(
        stats: VehicleInventoryStats,
        vehicle: Vehicle
    ): List<InventoryAlert> {
        val alerts = mutableListOf<InventoryAlert>()
        val now = Date()

        if (stats.daysInInventory >= THRESHOLD_AGING_90) {
            alerts.add(
                InventoryAlert(
                    id = UUID.randomUUID(),
                    vehicleId = vehicle.id,
                    alertType = InventoryAlertType.aging_90_days,
                    severity = "high",
                    message = "Vehicle has been in inventory for ${stats.daysInInventory} days. Consider aggressive pricing.",
                    isRead = false,
                    createdAt = now,
                    dismissedAt = null
                )
            )
        } else if (stats.daysInInventory >= THRESHOLD_AGING_60) {
            alerts.add(
                InventoryAlert(
                    id = UUID.randomUUID(),
                    vehicleId = vehicle.id,
                    alertType = InventoryAlertType.aging_60_days,
                    severity = "medium",
                    message = "Vehicle has been in inventory for ${stats.daysInInventory} days. Review pricing strategy.",
                    isRead = false,
                    createdAt = now,
                    dismissedAt = null
                )
            )
        }

        stats.roiPercent?.let { roi ->
            if (roi.compareTo(BigDecimal(THRESHOLD_LOW_ROI)) < 0) {
                alerts.add(
                    InventoryAlert(
                        id = UUID.randomUUID(),
                        vehicleId = vehicle.id,
                        alertType = InventoryAlertType.low_roi,
                        severity = "high",
                        message = "Projected ROI is ${roi.setScale(1, RoundingMode.HALF_UP)}%. Consider cost reduction or price increase.",
                        isRead = false,
                        createdAt = now,
                        dismissedAt = null
                    )
                )
            }
        }

        val holdingCostPercentage = VehicleFinancialsCalculator.calculateHoldingCostPercentage(
            stats.holdingCostAccumulated,
            stats.totalCost
        )

        holdingCostPercentage?.let { percentage ->
            if (percentage.compareTo(BigDecimal(THRESHOLD_HIGH_HOLDING_COST_PERCENT)) > 0) {
                alerts.add(
                    InventoryAlert(
                        id = UUID.randomUUID(),
                        vehicleId = vehicle.id,
                        alertType = InventoryAlertType.high_holding_cost,
                        severity = "medium",
                        message = "Holding cost is ${percentage.setScale(1, RoundingMode.HALF_UP)}% of total cost. Consider faster turnover.",
                        isRead = false,
                        createdAt = now,
                        dismissedAt = null
                    )
                )
            }
        }

        if (shouldRecommendPriceReduction(stats, vehicle)) {
            alerts.add(
                InventoryAlert(
                    id = UUID.randomUUID(),
                    vehicleId = vehicle.id,
                    alertType = InventoryAlertType.price_reduction_needed,
                    severity = "medium",
                    message = "Based on aging and market conditions, consider a price reduction to accelerate sale.",
                    isRead = false,
                    createdAt = now,
                    dismissedAt = null
                )
            )
        }

        return alerts
    }

    fun calculateInventoryHealthScore(
        vehicles: List<Vehicle>,
        allStats: List<VehicleInventoryStats>
    ): Int {
        if (vehicles.isEmpty()) {
            return 100
        }

        var totalScore = 0

        allStats.forEach { stats ->
            var vehicleScore = 100

            when (stats.agingBucket) {
                AGING_BUCKET_31_60 -> vehicleScore -= 10
                AGING_BUCKET_61_90 -> vehicleScore -= 25
                AGING_BUCKET_90_PLUS -> vehicleScore -= 40
                else -> {}
            }

            stats.roiPercent?.let { roi ->
                when {
                    roi.compareTo(BigDecimal.ZERO) < 0 -> vehicleScore -= 30
                    roi.compareTo(BigDecimal("10")) < 0 -> vehicleScore -= 15
                    roi.compareTo(BigDecimal("20")) < 0 -> vehicleScore -= 5
                    else -> vehicleScore += 5
                }
            }

            val holdingCostPercentage = stats.holdingCostAccumulated
                .divide(stats.totalCost, 4, RoundingMode.HALF_UP)
                .multiply(BigDecimal(100))

            when {
                holdingCostPercentage.compareTo(BigDecimal("20")) > 0 -> vehicleScore -= 20
                holdingCostPercentage.compareTo(BigDecimal("10")) > 0 -> vehicleScore -= 10
            }

            totalScore += maxOf(0, minOf(100, vehicleScore))
        }

        return totalScore / vehicles.size
    }

    fun calculateAgingDistribution(
        stats: List<VehicleInventoryStats>
    ): Map<String, Int> {
        return stats.groupingBy { it.agingBucket }
            .eachCount()
            .toSortedMap(compareBy { bucket ->
                when (bucket) {
                    AGING_BUCKET_0_30 -> 0
                    AGING_BUCKET_31_60 -> 1
                    AGING_BUCKET_61_90 -> 2
                    AGING_BUCKET_90_PLUS -> 3
                    else -> 4
                }
            })
    }

    fun calculateAverageDaysInInventory(stats: List<VehicleInventoryStats>): Int {
        if (stats.isEmpty()) {
            return 0
        }
        return stats.sumOf { it.daysInInventory } / stats.size
    }

    fun calculateTotalHoldingCost(stats: List<VehicleInventoryStats>): BigDecimal {
        return stats
            .map { it.holdingCostAccumulated }
            .fold(BigDecimal.ZERO) { acc, cost -> acc.add(cost) }
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    fun calculateTotalInventoryValue(stats: List<VehicleInventoryStats>): BigDecimal {
        return stats
            .map { it.totalCost }
            .fold(BigDecimal.ZERO) { acc, cost -> acc.add(cost) }
            .setScale(DISPLAY_SCALE, RoundingMode.HALF_UP)
    }

    private fun shouldRecommendPriceReduction(
        stats: VehicleInventoryStats,
        vehicle: Vehicle
    ): Boolean {
        if (vehicle.saleDate != null) {
            return false
        }

        val isAging = stats.daysInInventory >= THRESHOLD_AGING_60

        val hasLowROI = stats.roiPercent?.let { roi ->
            roi.compareTo(BigDecimal("15")) < 0
        } ?: false

        val hasHighHoldingCost = stats.holdingCostAccumulated
            .divide(stats.totalCost, 4, RoundingMode.HALF_UP)
            .multiply(BigDecimal(100))
            .compareTo(BigDecimal("12")) > 0

        return isAging && (hasLowROI || hasHighHoldingCost)
    }
}

File: iOS Car Dealer Tracker/Ezcar24Business/Services/Calculators/InventoryMetricsCalculator.swift
import Foundation
import CoreData

class InventoryMetricsCalculator {
    
    private static let displayScale = 2
    
    private static let thresholdAging60 = 60
    private static let thresholdAging90 = 90
    private static let thresholdLowROI: Decimal = 10
    private static let thresholdHighHoldingCostPercent: Decimal = 15
    
    static func getAgingBucket(daysInInventory: Int) -> AgingBucket {
        return AgingBucket.fromDays(daysInInventory)
    }
    
    static func calculateHealthScore(
        vehicles: [Vehicle],
        stats: [UUID: VehicleInventoryStats]
    ) -> Int {
        guard !vehicles.isEmpty else {
            return 100
        }
        
        var totalScore = 0
        
        for vehicle in vehicles {
            guard let vehicleStats = stats[vehicle.id ?? UUID()] else {
                totalScore += 100
                continue
            }
            
            var vehicleScore = 100
            
            let agingBucket = AgingBucket.fromDays(Int(vehicleStats.daysInInventory))
            switch agingBucket {
            case .aging:
                vehicleScore -= 10
            case .stale:
                vehicleScore -= 25
            case .critical:
                vehicleScore -= 40
            default:
                break
            }
            
            if let roi = vehicleStats.roiPercent?.decimalValue {
                if roi < 0 {
                    vehicleScore -= 30
                } else if roi < 10 {
                    vehicleScore -= 15
                } else if roi < 20 {
                    vehicleScore -= 5
                } else {
                    vehicleScore += 5
                }
            }
            
            let holdingCostPercentage = (vehicleStats.holdingCostAccumulated?.decimalValue ?? 0) /
                (vehicleStats.totalCost?.decimalValue ?? 1) * 100
            
            if holdingCostPercentage > 20 {
                vehicleScore -= 20
            } else if holdingCostPercentage > 10 {
                vehicleScore -= 10
            }
            
            totalScore += max(0, min(100, vehicleScore))
        }
        
        return totalScore / vehicles.count
    }
    
    static func calculateTurnoverRatio(
        averageDaysInInventory: Int
    ) -> Double {
        guard averageDaysInInventory > 0 else {
            return 0.0
        }
        return Double(365) / Double(averageDaysInInventory)
    }
    
    static func calculateAgingDistribution(
        stats: [VehicleInventoryStats]
    ) -> [AgingBucket: Int] {
        var distribution: [AgingBucket: Int] = [:]
        
        for stat in stats {
            let bucket = AgingBucket.fromDays(Int(stat.daysInInventory))
            distribution[bucket, default: 0] += 1
        }
        
        return distribution
    }
    
    static func shouldGenerateAlert(
        vehicle: Vehicle,
        stats: VehicleInventoryStats
    ) -> InventoryAlertType? {
        let daysInInventory = Int(stats.daysInInventory)
        
        if daysInInventory >= thresholdAging90 {
            return .longDaysInInventory
        }
        
        if daysInInventory >= thresholdAging60 {
            return .aging
        }
        
        if let roi = stats.roiPercent?.decimalValue, roi < thresholdLowROI {
            return .lowROI
        }
        
        let holdingCostPercentage = (stats.holdingCostAccumulated?.decimalValue ?? 0) /
            (stats.totalCost?.decimalValue ?? 1) * 100
        
        if holdingCostPercentage > thresholdHighHoldingCostPercent {
            return .highHoldingCost
        }
        
        if shouldRecommendPriceReduction(stats: stats, vehicle: vehicle) {
            return .priceDrop
        }
        
        return nil
    }
    
    static func calculateInventoryStats(
        vehicle: Vehicle,
        expenses: [Expense],
        settings: HoldingCostSettings,
        context: NSManagedObjectContext
    ) -> VehicleInventoryStats {
        let daysInInventory = HoldingCostCalculator.calculateDaysInInventory(vehicle: vehicle)
        let agingBucket = getAgingBucket(daysInInventory: daysInInventory)
        
        let holdingCostAccumulated = HoldingCostCalculator.calculateAccumulatedHoldingCost(
            vehicle: vehicle,
            settings: settings,
            allExpenses: expenses
        )
        
        let totalCost = VehicleFinancialsCalculator.calculateTotalCost(
            vehicle: vehicle,
            expenses: expenses,
            holdingCost: holdingCostAccumulated
        )
        
        let roiPercent: NSDecimalNumber? = vehicle.salePrice.flatMap { salePrice in
            let roi = VehicleFinancialsCalculator.calculateROI(
                salePrice: salePrice.decimalValue,
                totalCost: totalCost
            )
            return roi.map { NSDecimalNumber(decimal: $0) }
        }
        
        let profitEstimate: NSDecimalNumber? = vehicle.askingPrice.flatMap { askingPrice in
            let profit = VehicleFinancialsCalculator.calculateProfitEstimate(
                askingPrice: askingPrice.decimalValue,
                totalCost: totalCost
            )
            return profit.map { NSDecimalNumber(decimal: $0) }
        }
        
        let now = Date()
        
        let stats = VehicleInventoryStats(context: context)
        stats.id = UUID()
        stats.vehicleId = vehicle.id
        stats.daysInInventory = Int32(daysInInventory)
        stats.agingBucket = agingBucket.rawValue
        stats.totalCost = NSDecimalNumber(decimal: totalCost)
        stats.holdingCostAccumulated = NSDecimalNumber(decimal: holdingCostAccumulated)
        stats.roiPercent = roiPercent
        stats.profitEstimate = profitEstimate
        stats.lastCalculatedAt = now
        stats.createdAt = now
        stats.updatedAt = now
        
        return stats
    }
    
    static func generateInventoryAlerts(
        stats: VehicleInventoryStats,
        vehicle: Vehicle,
        context: NSManagedObjectContext
    ) -> [InventoryAlert] {
        var alerts: [InventoryAlert] = []
        let now = Date()
        
        let daysInInventory = Int(stats.daysInInventory)
        
        if daysInInventory >= thresholdAging90 {
            let alert = InventoryAlert(context: context)
            alert.id = UUID()
            alert.vehicleId = vehicle.id
            alert.alertType = InventoryAlertType.longDaysInInventory.rawValue
            alert.severity = "high"
            alert.message = "Vehicle has been in inventory for \(daysInInventory) days. Consider aggressive pricing."
            alert.isRead = false
            alert.createdAt = now
            alert.dismissedAt = nil
            alerts.append(alert)
        } else if daysInInventory >= thresholdAging60 {
            let alert = InventoryAlert(context: context)
            alert.id = UUID()
            alert.vehicleId = vehicle.id
            alert.alertType = InventoryAlertType.aging.rawValue
            alert.severity = "medium"
            alert.message = "Vehicle has been in inventory for \(daysInInventory) days. Review pricing strategy."
            alert.isRead = false
            alert.createdAt = now
            alert.dismissedAt = nil
            alerts.append(alert)
        }
        
        if let roi = stats.roiPercent?.decimalValue, roi < thresholdLowROI {
            let alert = InventoryAlert(context: context)
            alert.id = UUID()
            alert.vehicleId = vehicle.id
            alert.alertType = InventoryAlertType.lowROI.rawValue
            alert.severity = "high"
            alert.message = String(format: "Projected ROI is %.1f%%. Consider cost reduction or price increase.", NSDecimalNumber(decimal: roi).doubleValue)
            alert.isRead = false
            alert.createdAt = now
            alert.dismissedAt = nil
            alerts.append(alert)
        }
        
        let holdingCostPercentage = (stats.holdingCostAccumulated?.decimalValue ?? 0) /
            (stats.totalCost?.decimalValue ?? 1) * 100
        
        if holdingCostPercentage > thresholdHighHoldingCostPercent {
            let alert = InventoryAlert(context: context)
            alert.id = UUID()
            alert.vehicleId = vehicle.id
            alert.alertType = InventoryAlertType.highHoldingCost.rawValue
            alert.severity = "medium"
            alert.message = String(format: "Holding cost is %.1f%% of total cost. Consider faster turnover.", NSDecimalNumber(decimal: holdingCostPercentage).doubleValue)
            alert.isRead = false
            alert.createdAt = now
            alert.dismissedAt = nil
            alerts.append(alert)
        }
        
        if shouldRecommendPriceReduction(stats: stats, vehicle: vehicle) {
            let alert = InventoryAlert(context: context)
            alert.id = UUID()
            alert.vehicleId = vehicle.id
            alert.alertType = InventoryAlertType.priceDrop.rawValue
            alert.severity = "medium"
            alert.message = "Based on aging and market conditions, consider a price reduction to accelerate sale."
            alert.isRead = false
            alert.createdAt = now
            alert.dismissedAt = nil
            alerts.append(alert)
        }
        
        return alerts
    }
    
    static func calculateAverageDaysInInventory(
        stats: [VehicleInventoryStats]
    ) -> Int {
        guard !stats.isEmpty else {
            return 0
        }
        let totalDays = stats.reduce(0) { $0 + Int($1.daysInInventory) }
        return totalDays / stats.count
    }
    
    static func calculateTotalHoldingCost(
        stats: [VehicleInventoryStats]
    ) -> Decimal {
        return stats
            .reduce(Decimal(0)) { $0 + ($1.holdingCostAccumulated?.decimalValue ?? 0) }
            .rounded(to: displayScale)
    }
    
    static func calculateTotalInventoryValue(
        stats: [VehicleInventoryStats]
    ) -> Decimal {
        return stats
            .reduce(Decimal(0)) { $0 + ($1.totalCost?.decimalValue ?? 0) }
            .rounded(to: displayScale)
    }
    
    private static func shouldRecommendPriceReduction(
        stats: VehicleInventoryStats,
        vehicle: Vehicle
    ) -> Bool {
        guard vehicle.saleDate == nil else {
            return false
        }
        
        let isAging = stats.daysInInventory >= thresholdAging60
        
        let hasLowROI = stats.roiPercent.flatMap { roi in
            roi.decimalValue < 15
        } ?? false
        
        let holdingCostPercentage = (stats.holdingCostAccumulated?.decimalValue ?? 0) /
            (stats.totalCost?.decimalValue ?? 1) * 100
        let hasHighHoldingCost = holdingCostPercentage > 12
        
        return isAging && (hasLowROI || hasHighHoldingCost)
    }
}

private extension Decimal {
    func rounded(to scale: Int) -> Decimal {
        var result = Decimal()
        var localCopy = self
        NSDecimalRound(&result, &localCopy, scale, .plain)
        return result
    }
}
